src/finality/grandpa/warp_sync.rs

```diff
@@ -0,0 +1,78 @@
+use crate::chain::chain_information::{ChainInformation, ChainInformationFinality};
+use crate::finality::justification::verify::{
+    verify, Config as VerifyConfig, Error as VerifyError,
+};
+use crate::header::{DigestItemRef, GrandpaConsensusLogRef};
+use crate::network::service::GrandpaWarpSyncResponseFragment;
+
+pub struct Verifier {
+    index: usize,
+    authorities_list: Vec<[u8; 32]>,
+    fragments: Vec<GrandpaWarpSyncResponseFragment>,
+}
+
+impl Verifier {
+    pub fn new(
+        genesis_chain_infomation: &ChainInformation,
+        warp_sync_response_fragments: Vec<GrandpaWarpSyncResponseFragment>,
+    ) -> Self {
+        let authorities_list = match &genesis_chain_infomation.finality {
+            ChainInformationFinality::Grandpa {
+                finalized_triggered_authorities,
+                ..
+            } => finalized_triggered_authorities
+                .iter()
+                .map(|auth| auth.public_key)
+                .collect(),
+            _ => unimplemented!(),
+        };
+
+        Self {
+            index: 0,
+            authorities_list,
+            fragments: warp_sync_response_fragments,
+        }
+    }
+
+    pub fn next(mut self) -> Result<Next, VerifyError> {
+        let authorities_set_id = self.index as u64;
```

Comment:
```suggestion
        let authorities_set_id = u64::try_from(self.index).unwrap();
```

That's a coding guideline that I forgot to write out explicitly